### PR TITLE
DM-3368

### PR DESCRIPTION
**Pull request by r-owen**

### CODE REVIEW

File: python/lsst/pipe/drivers/singleFrameDriver.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
+import sys
```
Comment (r-owen) [inline python/lsst/pipe/drivers/singleFrameDriver.py]: This file has many linter warnings (unused imports, setting variables that are never used). Please fix those.

File: python/lsst/pipe/drivers/singleFrameDriver.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        """Constructor.
+
+        All nodes execute this method.
+        """
+        super(SingleFrameDriverTask, self).__init__(*args, **kwargs)
+        self.makeSubtask("processCcd")
+
+    @classmethod
+    def batchWallTime(cls, time, parsedCmd, numCpus):
+        numCcds = sum(1 for raft in parsedCmd.butler.get("camera") for ccd in raft)
```
Comment (r-owen) [inline python/lsst/pipe/drivers/singleFrameDriver.py:52]: Please add a doc string

File: python/lsst/pipe/drivers/singleFrameDriver.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        numCycles = int(math.ceil(numCcds/float(numCpus)))
+        numExps = len(cls.RunnerClass.getTargetList(parsedCmd))
+        return time*numExps*numCycles
+
+    @classmethod
+    def _makeArgumentParser(cls, *args, **kwargs):
+        doBatch = kwargs.pop("doBatch", False)
```
Comment (r-owen) [inline python/lsst/pipe/drivers/singleFrameDriver.py]: Why is this line needed? Please add a comment or doc string explain why an override is needed. Also, if this line is the only override, consider calling the parent class's implementation after this initial line to reduce code duplication.

File: python/lsst/pipe/drivers/singleFrameDriver.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+            self.log.warn("All CCDs in exposure failed")
+            return
+
+    def process(self, cache, dataId):
+        """Process a single CCD and save the results for a later write.
+
+        Only slaves execute this method.
+        """
+        cache.result = None
+        if dataId["ccd"] in self.config.ignoreCcdList:
```
Comment (r-owen) [inline python/lsst/pipe/drivers/singleFrameDriver.py]: I suggest you save the ignore list as a set in **init** and then check for existence in the set to speed search time.

File: python/lsst/pipe/drivers/utils.py [hunk 5]
```diff
@@ -1,7 +1,15 @@
-from lsst.pipe.base import Struct
+from lsst.pipe.base import Struct, TaskRunner
 from lsst.pipe.tasks.coaddBase import CoaddDataIdContainer
 from lsst.pipe.tasks.selectImages import BaseSelectImagesTask, BaseExposureInfo
 
+class ButlerTaskRunner(TaskRunner):
```
Comment (r-owen) [inline python/lsst/pipe/drivers/utils.py:8]: Consider adding this to pipe_base, as it might be more generally useful. That said, we have a lot of task runners that should probably go in pipe_base and it may not be worth the hassle to start down that road right now.

File: python/lsst/pipe/drivers/singleFrameDriver.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        parser.add_id_argument("--id", datasetType="raw", level="visit",
+                               help="data ID, e.g. --id visit=12345")
+        return parser
+
+    @abortOnError
+    def run(self, expRef, butler):
+        """Process a single exposure, with scatter-gather-scatter using MPI.
+
+        All nodes execute this method, though the master and slaves have different
+        routes through it.  The expRef is only a DummyDataRef on the slaves.
```
Comment (r-owen) [inline python/lsst/pipe/drivers/singleFrameDriver.py]: I could not find the test for DummyDataRef. It may be worth giving a hint to the reader as to where this occurs.

File: python/lsst/pipe/drivers/singleFrameDriver.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+        dataIdList = dict([(ccdRef.get("ccdExposureId"), ccdRef.dataId)
+                           for ccdRef in expRef.subItems("ccd") if ccdRef.datasetExists("raw")])
+        dataIdList = collections.OrderedDict(sorted(dataIdList.items()))
+
+        # Scatter: process CCDs independently
+        structList = pool.map(self.process, dataIdList.values())
+        numGood = sum(1 for s in structList if s is not None)
+        if numGood == 0:
+            self.log.warn("All CCDs in exposure failed")
```
Comment (r-owen) [inline python/lsst/pipe/drivers/singleFrameDriver.py:82]: Is it worth raising an exception for this case or at least logging an error?

File: python/lsst/pipe/drivers/singleFrameDriver.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+            return
+
+    def process(self, cache, dataId):
+        """Process a single CCD and save the results for a later write.
```
Comment (r-owen) [inline python/lsst/pipe/drivers/singleFrameDriver.py:86]: Please document the inputs and returned values

File: python/lsst/pipe/drivers/utils.py [hunk 9]
```diff
@@ -1,7 +1,15 @@
-from lsst.pipe.base import Struct
```
Comment (r-owen) [inline python/lsst/pipe/drivers/utils.py:1]: This file has some linter warnings and even one error (undefined name "argparse"). Please fix before merging to master.

Preferably run all files in this package through a linter (flake8 with suitable overrides, or pylint) before merging.

File: python/lsst/pipe/drivers/singleFrameDriver.py (hunk 4 above)
Comment (hsinfang) [inline python/lsst/pipe/drivers/singleFrameDriver.py]: Sorry for commenting while I'm not the reviewer. 

Would there be a way not to hard code the dataId key "ccd"? This wouldn't work with the current obs_decam where "ccdnum" is used instead of "ccd".

Comment (timj) [inline python/lsst/pipe/drivers/singleFrameDriver.py]: There is no problem with other people commenting.

Comment (r-owen) [inline python/lsst/pipe/drivers/singleFrameDriver.py]: I agree. Additional comments are always welcome; you may see something the reviewer missed, as in this case. I am dismayed by DECam using "ccdnum" and not having "ccd" available, but given this unusual usage a config parameter to specify the field name is probably in order.

Comment (TallJimbo) [inline python/lsst/pipe/drivers/singleFrameDriver.py]: I believe this would also fail with obs_lsstSim, which is "visit", "raft", "sensor".

I'd ultimately like to require that all mosaic cameras support some standard name for a single sensor for cases exactly like this, but that's not something we have right now (or even something I have time to cook up a fully-fledged proposal for).

Comment (r-owen) [inline python/lsst/pipe/drivers/singleFrameDriver.py]: Good point. Worse, LSSTSim IDs have separate "raft" and "sensor" standard keywords, and you need both to uniquely identify a CCD and "ccd" may be present as an alias for "sensor". LSSTSim IDs may have "ccdName" or "sensorName", either of which would suffice, but the present version of LsstSimMapper._transformId does not enforce their presence, so we cannot rely on them.

So even adding the name of a key to check would not suffice for LSSTSim. However, it would work for enough cases that it is probably worth doing. LSSTSim can adapt by always creating "sensorName" or by subclassing the task.

File: python/lsst/pipe/drivers/singleFrameDriver.py (hunk 6 above)
Comment (natelust) [inline python/lsst/pipe/drivers/singleFrameDriver.py]: It was a bit rotted string, removed now.
